src/poetry/packages/locker.py

```diff
@@ -608,6 +608,9 @@ def _dump_package(self, package: Package) -> dict[str, Any]:
             if package.source_resolved_reference:
                 data["source"]["resolved_reference"] = package.source_resolved_reference
 
+            if package.source_subdirectory:
+                data["source"]["subdirectory"] = package.source_subdirectory
```

Comment:
Looks like this also needs to be read back out of the lock file in `Locker.locked_repository`:

```diff
             package = Package(
                 info["name"],
                 info["version"],
                 info["version"],
                 source_type=source_type,
                 source_url=url,
                 source_reference=source.get("reference"),
                 source_resolved_reference=source.get("resolved_reference"),
+                source_subdirectory=source.get("subdirectory")
             )
```